File: weather/weather.go
package weather

import (
    "os"
    "fmt"
    "encoding/json"
    "io/ioutil"
    "net/http"
)

// OpenWeather api (https://openweathermap.org/current) を使って天気情報を取得
func GetWeather() []Result {
    // GET リクエストを投げる
    apiKey := os.Getenv("OPEN_WEATHER_MAP_API_KEY")
    cityName := "Yokohama"
    url := fmt.Sprintf("http://api.openweathermap.org/data/2.5/forecast?q=%s&appid=%s&lang=ja&units=metric", cityName, apiKey)
    res, err := http.Get(url)
    if err != nil {
        panic(err)
    }

    defer res.Body.Close() 

    // レスポンス表示
    body, err := ioutil.ReadAll(res.Body)
    if err != nil{
        panic(err)
    }

    // jsonを構造体へデコード
    var data Response
    if err := json.Unmarshal(body, &data); err != nil {
        panic(err)
    }
    
    var result []Result 
    for _, d := range data.List {
        var r Result
        r.Date = d.DtTxt
        r.Temp = d.Main.Temp
        r.Weather = d.Weather[0].Description
        result = append(result, r)
    }

    // fmt.Printf("%v", result)
    return result
}

// getWeather関数の戻り値
type Result struct {
    Date	string	`json:"date"`
    Temp	float64	`json:"temp"`
    Weather	string	`json:"weather"` 
}

// 自動生成 by https://mholt.github.io/json-to-go/
type Response struct {
    Cod     string `json:"cod"`
    Message int    `json:"message"`
    Cnt     int    `json:"cnt"`
    List    []struct {
        Dt   int `json:"dt"`
        Main struct {
            Temp      float64 `json:"temp"`
            FeelsLike float64 `json:"feels_like"`
            TempMin   float64 `json:"temp_min"`
            TempMax   float64 `json:"temp_max"`
            Pressure  int     `json:"pressure"`
            SeaLevel  int     `json:"sea_level"`
            GrndLevel int     `json:"grnd_level"`
            Humidity  int     `json:"humidity"`
            TempKf    float64 `json:"temp_kf"`
        } `json:"main"`
        Weather []struct {
            ID          int    `json:"id"`
            Main        string `json:"main"`
            Description string `json:"description"`
            Icon        string `json:"icon"`
        } `json:"weather"`
        Clouds struct {
            All int `json:"all"`
        } `json:"clouds"`
        Wind struct {
            Speed float64 `json:"speed"`
            Deg   int     `json:"deg"`
            Gust  float64 `json:"gust"`
        } `json:"wind"`
        Visibility int     `json:"visibility"`
        Pop        float64 `json:"pop"`
        Rain       struct {
            ThreeH float64 `json:"3h"`
        } `json:"rain,omitempty"`
        Sys struct {
            Pod string `json:"pod"`
        } `json:"sys"`
        DtTxt string `json:"dt_txt"`
    } `json:"list"`
    City struct {
        ID    int    `json:"id"`
        Name  string `json:"name"`
        Coord struct {
            Lat float64 `json:"lat"`
            Lon float64 `json:"lon"`
        } `json:"coord"`
        Country    string `json:"country"`
        Population int    `json:"population"`
        Timezone   int    `json:"timezone"`
        Sunrise    int    `json:"sunrise"`
        Sunset     int    `json:"sunset"`
    } `json:"city"`
}

File: src/functions/weather.py

import config
import requests

def get_weather():
    '''
    OpenWeather api (https://openweathermap.org/current) を使って天気情報を取得
    '''
    # リクエストの作成
    url = 'http://api.openweathermap.org/data/2.5/forecast'
    params = {
        'q': 'Yokohama', # city name
        'appid': config.OPEN_WEATHER_MAP_API_KEY,
        'lang': 'ja',
        'units': 'metric',
    }

    # GET リクエストを送信し、html を取得
    data = requests.get(url, params).json()
    
    '''
    data の中身 (See. https://openweathermap.org/current#current_JSON)
    {
        'cod': '200',
        'message': 0, 
        'cnt': 40,
        'list': [{
            'dt': 1628575200,
            'main': {
                'temp': 35.77,
                'feels_like': 42.77,
                'temp_min': 33.1,
                'temp_max': 35.77, 
                'pressure': 998, 
                'sea_level': 998, 
                'grnd_level': 998,
                'humidity': 51, 
                'temp_kf': 2.67
            },
            'weather': [{
                'id':801, 
                'main': 'Clouds', 
                'description': '薄い雲',
                'icon': '02d'
            }], 
            'clouds': {'all': 20}, 
            'wind': {'speed': 4.57, 'deg': 234, 'gust': 7.9}, 
            'visibility': 10000, 
            'pop': 0,
            'sys': {'pod': 'd'}, 
            'dt_txt': '2021-08-10 06:00:00'
        }, ... ], 
        'city': {
            'id': 1848354, 
            'name': '横浜市', 
            'coord': {'lat': 35.4478, 'lon': 139.6425}, 
            'country': 'JP', 
            'population': 3574443, 
            'timezone': 32400, 
            'sunrise': 1628538999, 
            'sunset': 1628588228
        }
    }
    '''

    # データ整形
    '''
    整形後
    {  
        'date1': {
            'time1': { temp, weather }, 
            'time2': { temp, weather }, 
            ...,
        },
        'date2': {
            'time1': { temp, weather }, 
            'time2': { temp, weather }, 
            ...,
        },
        ...,
    }
    '''
    formated_data = {}
    for d in data['list']:
        # 日付と時刻の取得
        date, time = d['dt_txt'].split(' ')

        # 日付の変換 2021-01-01 -> 2021年1月1日
        year, month, day = date.split('-')
        date = str(year) + '年' + str(int(month)) + '月' + str(int(day)) + '日'

        # 時刻の変換 09:00:00 -> 9時
        time = str(int(time[:2])) + '時'

        # 日付毎にdictを作成
        if not date in formated_data:
            formated_data[date] = {}

        formated_data[date][time] = {
            'temp': d['main']['temp'],
            'weather': d['weather'][0]['description']
        }
    
    return {'city': data['city']['name'], 'data': formated_data}